src/kernel.hpp

```diff
@@ -212,11 +212,17 @@ struct cvk_kernel_argument_values {
     cl_int set_arg(const kernel_argument& arg, size_t size, const void* value) {
 
         if (arg.is_pod()) {
-            if (size != arg.size) {
+            // If the argument is a vec3, OpenCL requires to call clSetKernelArg
+            // with a size of 4 times the element size. But clspv arg size is
+            // only 3 times the element size. When size and arg.size do not
+            // match, make sure that we are not in this unusual case.
+            if (size != arg.size &&
+                !(arg.info.type_name[arg.info.type_name.length() - 1] == '3' &&
```

Comment:
I suggest you introduce `kernel_argument::is_vec3()` to help with readability (don't forget to assert the validity of `kernel_argument_info::extended_valid` there) and separation of concerns.